lib/units/api/controllers/user.js

```diff
@@ -0,0 +1,332 @@
+            datautil.normalize(device, req.user)
+            if (fields) {
+              device = _.pick(device, fields.split(','))
+            }
+            deviceList.push(device)
+          })
+
+          res.json({
+            success: true
+          , devices: deviceList
+          })
+        })
+    })
+    .catch(function(err) {
+      log.error('Failed to load device list: ', err.stack)
+      res.status(500).json({
+        success: false
+      })
+    })
+}
+
+function getUserDeviceBySerial(req, res) {
+  var serial = req.swagger.params.serial.value
+  var fields = req.swagger.params.fields.value
+
+  dbapi.loadDevice(serial)
+    .then(function(device) {
+      if (device) {
+        datautil.normalize(device, req.user)
+
+        if (device.owner && device.owner.email === req.user.email) {
+          if(fields) {
+            device = _.pick(device, fields.split(','))
+          }
+
+          res.json({
+            success: true
+          , device: device
+          })
+        }
+        else {
+          res.status(401).json({
+            success: false
+          , description: 'Device is not owned by you'
+          })
+        }
+      }
+      else {
+        res.status(404).json({
+          success: false
+        , description: 'Device not found'
+        })
+      }
+    })
+    .catch(function(err) {
+      log.error('Failed to load device "%s": ', req.params.serial, err.stack)
+      res.status(500).json({
+        success: false
+      })
+    })
+}
+
+function addUserDevice(req, res) {
+  var serial = req.body.serial
+  var timeout = req.body.timeout || null
+
+  dbapi.loadDevice(serial)
+    .then(function(device) {
+      if (device) {
+        datautil.normalize(device, req.user)
+        if(device.present && device.ready && !device.using && !device.owner) {
+
+          var requirements = {
+            'serial': {
+              'value': serial,
+              'match': 'exact'
+            }
+          }
+
+          req.options.push.send([
+            device.channel
+          , wireutil.envelope(
+              new wire.GroupMessage(
+                new wire.OwnerMessage(
+                  req.user.email
+                , req.user.name
+                , req.user.group
+                )
+              , timeout
+              , wireutil.toDeviceRequirements(requirements)
+              )
+            )
+          ])
+
+          res.status(202).json({
+            success: true
+          , description: 'Device Add request is accepted. Check if device is successfully added using pollingUrl'
+          , pollingUrl: util.format('%s://%s%s/user/devices/%s'
+                                      , req.protocol
+                                      , req.get('host')
+                                      , req.swagger.operation.api.basePath
+                                      , serial
+                                    )
+          })
+
+        } else {
+          res.status(401).json({
```

Comment:
Technically you're not allowed to return a 401 without a WWW-Authenticate header.

> The response MUST include a WWW-Authenticate header field (section 14.47) containing a challenge applicable to the requested resource.

Which is very annoying, and basically makes it only work with HTTP auth unless you're willing to violate the spec. The OAuth spec uses something like `WWW-Authenticate: Bearer realm="example"`, though, so stealing that would also be an option.